pkg/experiment/metastore/blockcleaner/block_cleaner.go

```diff
@@ -0,0 +1,417 @@
+	cntDeleted := 0
+	cntDeletedBucket := 0
+	for blockId, removalContext := range blocks {
+		if removalContext.expiryTs < now {
+			metricLabels := []string{removalContext.tenant, fmt.Sprint(shard)}
+			if c.isLeader {
+				var key string
+				if removalContext.tenant != "" {
+					key = filepath.Join("blocks", fmt.Sprint(shard), removalContext.tenant, blockId, "block.bin")
+				} else {
+					key = filepath.Join("segments", fmt.Sprint(shard), "anonymous", blockId, "block.bin")
+				}
+				level.Debug(c.logger).Log(
+					"msg", "removing block from bucket",
+					"shard", shard,
+					"tenant", removalContext.tenant,
+					"blockId", blockId,
+					"expiryTs", removalContext.expiryTs,
+					"bucket_key", key)
+				err := c.bkt.Delete(ctx, key)
+				if err != nil {
+					level.Warn(c.logger).Log(
+						"msg", "failed to remove block from bucket",
+						"err", err,
+						"blockId", blockId,
+						"shard", shard,
+						"tenant", removalContext.tenant)
+					// TODO(aleks-p): Detect if the error is "object does not exist" or something else. Handle each case appropriately.
+					continue
+				}
+				c.metrics.bucketObjectRemovals.WithLabelValues(metricLabels...).Inc()
+				cntDeletedBucket++
+			}
+			err = c.removeBlock(blockId, shard, removalContext)
+			if err != nil {
+				level.Warn(c.logger).Log(
+					"msg", "failed to remove block from pending block removals",
+					"err", err,
+					"blockId", blockId,
+					"shard", shard,
+					"tenant", removalContext.tenant,
+					"expiry", removalContext.expiryTs)
+			}
+			level.Debug(c.logger).Log(
+				"msg", "removed block from pending block removals",
+				"blockId", blockId,
+				"shard", shard,
+				"tenant", removalContext.tenant,
+				"expiryTs", removalContext.expiryTs)
+			c.metrics.expiredBlocks.WithLabelValues(metricLabels...).Inc()
+			cntDeleted++
+		}
+	}
+	level.Info(c.logger).Log("msg", "finished shard cleanup", "shard", shard, "blocks_removed", cntDeleted, "blocks_removed_bucket", cntDeletedBucket)
+	return nil
+}
+
+func (c *blockCleaner) listBlocks(shard uint32) (map[string]*blockRemovalContext, error) {
+	blocks := make(map[string]*blockRemovalContext)
+	err := c.db().View(func(tx *bbolt.Tx) error {
+		bkt, err := getPendingBlockRemovalsBucket(tx)
+		if err != nil {
+			return err
+		}
+		shardBkt := bkt.Bucket(getShardBucketName(shard))
+		if shardBkt == nil {
+			return nil
+		}
+		return shardBkt.ForEach(func(k, v []byte) error {
+			if len(k) < 34 {
+				return fmt.Errorf("block key too short (expected 34 chars, was %d)", len(k))
+			}
+			blockId := string(k[:26])
+			blocks[blockId] = &blockRemovalContext{
+				expiryTs: int64(binary.BigEndian.Uint64(k[26:34])),
+				tenant:   string(k[34:]),
+			}
+			return nil
+		})
+	})
+	if err != nil {
+		return nil, err
+	}
+	return blocks, nil
+}
+
+func (c *blockCleaner) removeBlock(blockId string, shard uint32, removalContext *blockRemovalContext) error {
+	err := c.db().Update(func(tx *bbolt.Tx) error {
```

Comment:
This comment is not specific to this specific call, but concerns use of transactions in general: due to performance considerations we want to minimize the number of them and batch operations as much as we can